src/realm/sync/noinst/migration_store.cpp

```diff
@@ -35,13 +34,14 @@ MigrationStore::MigrationStore(DBRef db,
                                std::function<void(MigrationStore::MigrationState)>&& on_migration_state_changed)
     : m_db(std::move(db))
     , m_on_migration_state_changed(std::move(on_migration_state_changed))
+    , m_state(MigrationState::NotMigrated)
+    , m_query_string{}
 {
     std::vector<SyncMetadataTable> internal_tables{
```

Comment:
I propose that we move the code that actually opens the `SyncMetadataSchemaVersions` and sets up the tables to the `migrate_to_flx()` function so that all this stuff gets lazily loaded/constructed. I think that will 1) make it so realms that start out as FLX don't end up with this table with kinda invalid data in it and 2) reduce some of the side-effects where just opening any realm regardless of migration status causes side-effects.

Reply:
Updated `SyncMetadataSchemaVersions` to be created with a read_only param that will not initialize the data if it doesn't exist and added an `is_initialized()` function to check if the constructor with read_only = true was successful or not.

Reply:
nice change. I think it would be better to separate the read-only SyncMetadataSchemaVersions from the regular SyncMetadataSchemaVersions though. You could consider having a SyncMetadataSchemaVersionsReader which would be used by the MigrationStore constructor and the regular SyncMetadataSchemaVersions (it may lead to a bit of code duplication depending how it's implemented)

Reply:
Created separate `SyncMetadataSchemaVersionsReader` that only reads the data (with the same `is_initialized()` accessor in case the data does not exist) and added tests for both `SyncMetadataSchemaVersionsReader` and `SyncMetadataSchemaVersions` classes.